File: js/complete/no023.js
#!/usr/bin/env node

/* A number n is called abundant if the sum of its proper divisors exceeds n.

   By mathematical analysis, it can be shown that all integers greater than
   28123 can be written as the sum of two abundant numbers.

   Find the sum of all the positive integers which cannot be written as
   the sum of two abundant numbers. */

var fns = require('../functions.js'),
    operator = require('../operator.js'),
    timer = require('../timer.js');

function abundantNumbers(n) {
  var factorHash = fns.allFactors(n),
      // sum of proper divisors
      result = [];
  for (var i = 2; i <= n; i++) {
    if (i < operator.sum(factorHash[i]) - i) {
      result.push(i);
    }
  }
  return result;
};

// Array size limit is 4294967295 or 2^32 - 1
// http://4umi.com/web/javascript/array.php
exports.main = function(verbose) {
  if (typeof verbose == 'undefined') {
    verbose = false;
  }
  var abundants = abundantNumbers(28123);

  // Had to break from python implementation because of memory considerations,
  // as a result, the algo is better
  var sums = {}, i, val1, j, val2;
  for (var i = 0; i <= 28123; i++) {
    sums[i] = false;
  }

  for (i = 0; val1 = abundants[i]; i++) {
    for (j = i; val2 = abundants[j]; j++) {
        if (val1 + val2 <= 28123) {
          sums[val1 + val2] = true;
        }
    }
  }

  var badOnes = [];
  for (var i = 0; i <= 28123; i++) {
    if (!sums[i]) {
      badOnes.push(i);
    }
  }

  return operator.sum(badOnes);
};

if (require.main === module) {
  timer.timer(23, exports.main);
}

File: python/complete/no023.py
#!/usr/bin/env python

# A number n is called abundant if the sum of its proper divisors exceeds n.

# By mathematical analysis, it can be shown that all integers greater than
# 28123 can be written as the sum of two abundant numbers.

# Find the sum of all the positive integers which cannot be written as
# the sum of two abundant numbers.

import operator

from python.decorators import euler_timer
from python.functions import all_factors


def abundant_numbers(n):
    factor_hash = all_factors(n)
    # sum of proper divisors
    return [i for i in range(2, n + 1) if i < sum(factor_hash[i]) - i]


def main(verbose=False):
    abundants = abundant_numbers(28123)
    sums = [False] * (28123 + 1)

    length = len(abundants)
    for index in range(length):
        for second_index in range(index, length):
            val1 = abundants[index]
            val2 = abundants[second_index]
            if (val1 + val2 <= 28123):
                sums[val1 + val2] = True

    # those with indices set to false are the ones which can't be written
    # as the sum of two abundant numbers, so we sum them
    return sum(i for i, bool_val in enumerate(sums) if not bool_val)

if __name__ == '__main__':
    print euler_timer(23)(main)(verbose=True)
